apps/src/lab2/utils/getDeferredPromise.ts

```diff
@@ -0,0 +1,40 @@
+/*
+  a deferred promise is a promise which can be resolved at any time to hand arguments into a function.
+  e.g.,
+
+  async function f(deferred) {
+    return new Promise( (resolve, reject) => {
+      const args = await deferred
+      console.log("invoked with args: ", args)
+      // do interesting things with args
+    })
+  }
+
+  const { deferred, unlock } = getDeferredPromise()
+  const pendingPromise = f(deferred)
+  // you now have a pendingPromise. This can be handed into anything expecting a promise, even if you don't know
+  // the arguments you want to call it with yet.
+  At some point later, you can call:
+  resolve(["A","B", "C"])
+  and ["A", "B", "C"] will be handed into that pendingPromise and you can do what you want with it, then resolve at your leisure.
+}
+*/
+
+type PromiseFunction = ((value?: unknown) => void) | null;
+
+export type DeferredPromiseObject = {
+  deferred: Promise<unknown>;
+  resolve: PromiseFunction;
+  reject: PromiseFunction;
+};
+
+export function getDeferredPromise(): DeferredPromiseObject {
+  let resolve: PromiseFunction = null;
+  let reject: PromiseFunction = null;
+  const deferred = new Promise((inner_resolve, inner_reject) => {
+    resolve = inner_resolve;
+    reject = inner_reject;
+  });
+
+  return {deferred, resolve, reject};
+}
```

Comment:
This is the coolest bit of code in here. Normally we think of promises as things that we deal with in place - it's an action that takes place later, and then the promise tidies itself up and returns focus to the invoking code when it's done.

This one is a little different - the promise itself doesn't do _anything_. It just serves as a lock. I've called this structure a `padlock` in the past, but I know we tend to dislike whimsical names.

All this does is take the `promise`'s `resolve` and `reject` functions and copy them externally. Now we end up with a promise that we can hand around as desired _and_ we can have any arbitrary bit of code _outside_ the promise later wake up and resolve/reject it to return to the external context.

This is the secret sauce that lets the async dialogs work - `dialogControl.showDialog()` will return one of these deferred promises, so the programmer can either `await` it or process it with `then`/`catch`. But then when the dialog itself closes via the user hitting one of the buttons, that button press resolve the promise to finish it off and return control.

There is an additional bucket in the `DialogControlContext` to place args that will be given to the promise.

But this function itself is completely generic and can be used anywhere it's useful.